README.md
***WARNING: THIS REPO IS AN AUTO-GENERATED COPY.*** *This repo has been copied from [Gruntwork’s](https://gruntwork.io/) GitHub repositories so that you can consume it from your company’s own internal Git repositories. This copy is automatically created and updated by the `repo-copier` CLI tool. If you need to make changes to this repo, you should make the changes in a separate fork, and NOT make changes directly in this repo, as otherwise, the `repo-copier` will overwrite your changes! Please see the `repo-copier` [documentation](https://github.com/terraform-modules-krish/repo-copier) for more information on how the code is copied, how cross-references are updated, how the changelog is handled, etc.*

***

_You may find it valuable to view the following resources in the original repo. If these links give you a 404, visit https://app.gruntwork.io to gain access or email support@gruntwork.io if you need assistance._

[Home Page](https://github.com/gruntwork-io/gruntwork-installer/) |
[Pull Requests](https://github.com/gruntwork-io/gruntwork-installer/pulls) |
[Issues](https://github.com/gruntwork-io/gruntwork-installer/issues) |
[Releases and Assets](https://github.com/gruntwork-io/gruntwork-installer/releases)

_Alternatively, you can view a copied version of the resources listed above._

[Pull Requests](https://github.com/terraform-modules-krish/gruntwork-installer/blob/main/.github/PULL_REQUESTS.md) |
[Issues](https://github.com/terraform-modules-krish/gruntwork-installer/blob/main/.github/ISSUES.md) |
[ChangeLog](https://github.com/terraform-modules-krish/gruntwork-installer/blob/main/.github/CHANGELOG.md)

***

[![Maintained by Gruntwork.io](https://img.shields.io/badge/maintained%20by-gruntwork.io-%235849a6.svg)](https://gruntwork.io/?ref=repo_gruntwork_installer)
# Gruntwork Installer

`gruntwork-install` is a bash script you run to easily download and install "Gruntwork Modules".

## Compatibility

Tested under CentOS 7, latest Amazon Linux and Ubuntu 16.04.

## Quick Start

### Install gruntwork-install

If `gruntwork-install` is our approach for installing Gruntwork Modules, how do we install `gruntwork-install` itself?

Our solution is to make the `gruntwork-install` tool open source and to publish a `bootstrap-gruntwork-installer.sh`
script that anyone can use to install `gruntwork-install` itself. To use it, execute the following:

```
curl -LsS https://raw.githubusercontent.com/terraform-modules-krish/gruntwork-installer/master/bootstrap-gruntwork-installer.sh | bash /dev/stdin --version v0.0.22
```

Notice the `--version` parameter at the end where you specify which version of `gruntwork-install` to install. See the
[releases](https://github.com/gruntwork-io/gruntwork-installer/releases) page for all available versions.

For paranoid security folks, see [is it safe to pipe URLs into bash?](#is-it-safe-to-pipe-urls-into-bash) below.

### Use gruntwork-install

#### Authentication

To install scripts and binaries from private GitHub repos, you must create a [GitHub access
token](https://help.github.com/articles/creating-an-access-token-for-command-line-use/) and set it as the environment
variable `GITHUB_OAUTH_TOKEN` so `gruntwork-install` can use it to access the repo:

```
export GITHUB_OAUTH_TOKEN="(your secret token)"
```

#### Options

Once that environment variable is set, you can run `gruntwork-install` with the following options:

Option                      | Required | Description
--------------------------- | -------- | ------------
`--repo`                    | Yes      | The GitHub repo to install from.
`--tag`                     | Yes      | The version of the `--repo` to install from.<br>Follows the syntax described at [Tag Constraint Expressions](https://github.com/terraform-modules-krish/fetch#tag-constraint-expressions). This value is exposed to module install scripts as GRUNTWORK_INSTALL_TAG.
`--module-name`             | XOR      | The name of a module to install.<br>Can be any folder within the `modules` directory of `--repo`.<br>You must specify exactly one of `--module-name` or `--binary-name`.
`--binary-name`             | XOR      | The name of a binary to install.<br>Can be any file uploaded as a release asset in `--repo`.<br>You must specify exactly one of `--module-name` or `--binary-name`.
`--binary-sha256-checksum`  | No       | The SHA256 checksum of the binary specified by `--binary-name`. Should be exactly 64 characters..
`--binary-sha512-checksum`  | No       | The SHA512 checksum of the binary specified by `--binary-name`. Should be exactly 128 characters..
`--module-param`            | No       | A key-value pair of the format `key=value` you wish to pass to the module as a parameter. May be used multiple times. <br>Note: a `--` will automatically be appended to the `key` when your module is invoked<br>See the documentation for each module to find out what parameters it accepts.
`--download-dir`            | No       | The directory to which the module will be downloaded and from which it will be installed.
`--binary-install-dir`      | No       | The directory to which the binary will be installed. Only applies to binaries (not modules). Default: DEFAULT_BIN_DIR.
`--no-sudo`                 | No       | When true, don't use sudo to install the binary into the install directory. Only applies to binaries (not modules). Default: false.
`--branch`                  | No       | Download the latest commit from this branch in --repo. This is an alternative to --tag,<br>and is used only for testing. This value is exposed to module install scripts as GRUNTWORK_INSTALL_BRANCH.
`--ref`                     | No       | Download the latest commit from this ref in --repo. This is an alternative to --tag,<br>and is used only for testing. This value is exposed to module install scripts as GRUNTWORK_INSTALL_BRANCH.
`--help`                    | No       | Show the help text and exit.

#### Examples

##### Example 1: Download and Install a Script Module with No Parameters

Install the [ecs-scripts
module](https://github.com/terraform-modules-krish/terraform-aws-ecs/tree/master/modules/ecs-scripts) from the [terraform-aws-ecs
repo](https://github.com/terraform-modules-krish/terraform-aws-ecs), version `v0.0.1`:

```
gruntwork-install --module-name 'ecs-scripts' --repo 'https://github.com/terraform-modules-krish/terraform-aws-ecs' --tag 'v0.0.1'
```

##### Example 2: Download and Install a Script Module with Parameters

Install the [fail2ban
module](https://github.com/terraform-modules-krish/terraform-aws-security/tree/master/modules/fail2ban) from the [terraform-aws-security
repo](https://github.com/terraform-modules-krish/terraform-aws-security), passing two custom parameters to it:


```
gruntwork-install --module-name 'fail2ban' --repo 'terraform-aws-security' -module-param 'ban-time=3600'
```

##### Example 3: Download and Install a Binary Module

Install the `gruntkms` binary from the `v0.0.1` release of the [gruntkms
repo](https://github.com/terraform-modules-krish/gruntkms):

```
gruntwork-install --binary-name 'gruntkms' --repo 'https://github.com/terraform-modules-krish/gruntkms' --tag 'v0.0.1'
```

Note that the [v0.0.1 release of the gruntkms repo](https://github.com/gruntwork-io/gruntkms/releases/tag/v0.0.1) has
multiple binaries (`gruntkms_linux_amd64`, `gruntkms_darwin_386`, etc): `gruntwork-install` automatically picks the
right binary for your OS and copies it to `/usr/local/bin/gruntkms`.

##### Example 4: Use `gruntwork-install` in a Packer template

Finally, to put all the pieces together, here is an example of a Packer template that installs `gruntwork-install`
and then uses it to install several modules:

```json
{
  "variables": {
    "github_auth_token": "{{env `GITHUB_OAUTH_TOKEN`}}"
  },
  "builders": [
    {
      "ami_name": "gruntwork-install-example-{{isotime | clean_ami_name}}",
      "instance_type": "t2.micro",
      "region": "us-east-1",
      "type": "amazon-ebs",
      "source_ami": "ami-fce3c696",
      "ssh_username": "ubuntu"
    }
  ],
  "provisioners": [
    {
      "type": "shell",
      "inline":
        "curl -Ls https://raw.githubusercontent.com/terraform-modules-krish/gruntwork-installer/master/bootstrap-gruntwork-installer.sh | bash /dev/stdin --version v0.0.16"
    },
    {
      "type": "shell",
      "inline": [
        "gruntwork-install --module-name 'ecs-scripts' --repo 'https://github.com/terraform-modules-krish/terraform-aws-ecs' --tag 'v0.0.1'",
        "gruntwork-install --module-name 'fail2ban' --repo 'https://github.com/terraform-modules-krish/terraform-aws-security' -module-param 'ban-time=3600'",
        "gruntwork-install --binary-name 'gruntkms' --repo 'https://github.com/terraform-modules-krish/gruntkms' --tag 'v0.0.1'"
      ],
      "environment_vars": ["GITHUB_OAUTH_TOKEN={{user `github_auth_token`}}"]
    }
  ]
}
```

## Motivation

At [Gruntwork](http://www.gruntwork.io/), we've developed a number of scripts and binaries, most of them in private GitHub
repos, that perform common infrastructure tasks such as setting up continuous integration, monitoring, log aggregation,
and SSH access. Being able to use these "modules" of code typically involves many steps: you download the files
(possibly from a private GitHub repo), change their permissions, and run them with the parameters that make sense for
your environment.

That basically means lots of custom `bash` code copied differently across multiple software teams in multiple different
contexts. Worse, if we want to update a binary or script to add a new parameter, each team has to modify their own custom
code, which can be painful.

We believe we can do better by writing our scripts and binaries in a standardized way, and including a minimal tool that
streamlines the process of downloading and installing them. Also, since we give you 100% of the source code, we want it
to be clear exactly what happens when you install a Gruntwork Module.

Finally, installation should be streamlined no matter what platform (Windows, MacOS, Linux) you're on. Indeed, our goal
is to make installing Gruntwork Script Modules as easy as installing a typical package using `apt-get`, `yum`, `npm`,
or similar tools. We would have just used these existing tools, but none offer multi-platform compatibility.

## What's a Gruntwork Module?

A Gruntwork Module is a collection of one or more bash scripts and/or binaries maintained by Gruntwork that can be used to
add functionality to or configure an environment. There are multiple types of Gruntwork Modules:

* **Script Modules:** A collection of one or more files and scripts; installed with an `install.sh` script.
* **Binary Modules:** A single OS-specific executable binary.

Additional module types may be introduced in the future.

As an example, we have Script Modules for installing a CloudWatch Logs agent, optimizing syslog settings, and setting up
automatic security updates. We have a Binary Module for streamlining the use of Amazon Key Management Service (KMS).

Gruntwork sells [Infrastructure Packages](https://blog.gruntwork.io/gruntwork-infrastructure-packages-7434dc77d0b1#.6bwor6wxc).
Each Infrastructure Package corresponds to a specific GitHub repo and contains one or more Gruntwork Modules. The `/modules`
folder in the repo lists all Modules included with that Package.

### Freely Available Script Modules

Some Script Modules are so common that we've made them freely available in the [modules/](modules) folder of this repo.

### How `gruntwork-install` Works

To actually install a Gruntwork Module, we wrote a bash script named `gruntwork-install`. Here's how it works:

1. It uses [fetch](https://github.com/terraform-modules-krish/fetch) to download the specified version of the scripts or binary from
   the (public or private) git repo specified via the `--repo` option.
1. If you used the `--module-name` parameter, it downloads the files from the `modules` folder of `--repo` and runs
   the `install.sh` script of that module.
1. If you used the `--binary-name` parameter, it downloads the right binary for your OS, copies it to `/usr/local/bin`,
   and gives it execute permissions.

That's it!

## Create Your Own Gruntwork Modules

You can use `gruntwork-install` with any GitHub repo, not just repos maintained by Gruntwork.

That means that to create an installable Script Module, all you have to do is put it in the `modules` folder of
a GitHub repo to which you have access and include an `install.sh` script. To create a Binary Module, you just publish
it to a GitHub release with the name format `<NAME>_<OS>_<ARCH>`.

### Example

For example, in your Packer and Docker templates, you can use `gruntwork-install` to install the [ecs-scripts
module](https://github.com/terraform-modules-krish/terraform-aws-ecs/tree/master/modules/ecs-scripts) as follows:

```
gruntwork-install --module-name 'ecs-scripts' --repo 'https://github.com/terraform-modules-krish/terraform-aws-ecs' --tag 'v0.0.1'
```

In https://github.com/terraform-modules-krish/module-ecs, we download the contents of `/modules/ecs-scripts` and run
`/modules/ecs-scripts/install.sh`.

## Running tests

The tests for this repo are defined in the `test` folder. They are designed to run in a Docker container so that you
do not repeatedly dirty up your local OS while testing. We've defined a `test/docker-compose.yml` file as a convenient
way to expose the environment variables we need for testing and to mount local directories as volumes for rapid
iteration.

To run the tests:

1. Set your [GitHub access token](https://help.github.com/articles/creating-an-access-token-for-command-line-use/) as
   the environment variable `GITHUB_OAUTH_TOKEN`.
1. `./_ci/run-tests.sh`

## Security

### Validate the Downloaded Binary

Gruntwork-install will retrieve the desired GitHub Release Asset specified by the `--binary-name` property, but how can
we confirm that this binary has not been tampered with? In short, we trust that the maintainer has been responsible and
not allowed a malicious third-party to corrupt the Release Asset.

You can narrow the scope of this trust by computing a checksum on a Release Asset using a UNIX command like
`shasum -a 256 /path/to/file` when you first download the release. You can then feed this value (e.g. `b0b30cc24aed1b8cded2df903183b884c77f086efffc36ef19876d1c55fef93d`)
to `--binary-sha256-checksum` or `--binary-sha512-checksum`. If the checksum does not match, gruntwork-install will fail
with an error. This way, you are at least notified if the Release Asset you initially downloaded has since been changed.

### Is it safe to pipe URLs into bash?

Are you worried that our install instructions tell you to pipe a URL into bash? Although this approach has seen some
[backlash](https://news.ycombinator.com/item?id=6650987), we believe that the convenience of a one-line install
outweighs the minimal security risks. Below is a brief discussion of the most commonly discussed risks and what you can
do about them.

#### Risk #1: You don't know what the script is doing, so you shouldn't blindly execute it.

This is true of _all_ installers. For example, have you ever inspected the install code before running `apt-get install`
or `brew install` or double clicking a `.dmg` or `.exe` file? If anything, a shell script is the most transparent
installer out there, as it's one of the few that allows you to inspect the code (feel free to do so, as this script is
open source!). The reality is that you either trust the developer or you don't. And eventually, you automate the
install process anyway, at which point manual inspection isn't a possibility anyway.

#### Risk #2: The download URL could be hijacked for malicious code.

This is unlikely, as it is an https URL, and your download program (e.g. `curl`) should be verifying SSL certs. That
said, Certificate Authorities have been hacked in the past, and perhaps the Gruntwork GitHub account could be hacked
in the future, so if that is a major concern for you, feel free to copy the bootstrap code into your own codebase and
execute it from there. Alternatively, in the future we will publish checksums of all of our releases, so you could
optionally verify the checksum before executing the script.

#### Risk #3: The script may not download fully and executing it could cause errors.

We wrote our [bootstrap-gruntwork-installer.sh](bootstrap-gruntwork-installer.sh) as a series of bash functions that
are only executed by the very last line of the script. Therefore, if the script doesn't fully download, the worst
that'll happen when you execute it is a harmless syntax error.

## TODO

1. Add support for a `--version` flag to `bootstrap-gruntwork-installer.sh` and `gruntwork-install`.
1. Configure a CI build to automatically set the `--version` flag for each release.
1. Add an `uninstall` command that uses an `uninstall.sh` script in each module.
1. Add support for modules declaring their dependencies. Alternatively, consider Nix again as a dependency manager.


.github/CHANGELOG.md
# Changelog


# v0.0.39
_Released Apr 20, 2023_
## What's Changed
* Add GitHub PR & Issue Templates by @robmorgan in https://github.com/gruntwork-io/gruntwork-installer/pull/69
* Update PR Template by @rhoboat in https://github.com/gruntwork-io/gruntwork-installer/pull/74
* Update bootstrap-gruntwork-installer.sh by @iangrunt in https://github.com/gruntwork-io/gruntwork-installer/pull/75
* Update release version on README by @hongil0316 in https://github.com/gruntwork-io/gruntwork-installer/pull/76
* Update codeowners by @rhoboat in https://github.com/gruntwork-io/gruntwork-installer/pull/78
* Fix typo in README.md by @AlainODea in https://github.com/gruntwork-io/gruntwork-installer/pull/72
* Update deprecated circleci images to latest by @yorinasub17 in https://github.com/gruntwork-io/gruntwork-installer/pull/71
* Update old default branch references. by @rhoboat in https://github.com/gruntwork-io/gruntwork-installer/pull/81
* Update CODEOWNERS by @yorinasub17 in https://github.com/gruntwork-io/gruntwork-installer/pull/82
* Refactor contexts by @eak12913 in https://github.com/gruntwork-io/gruntwork-installer/pull/83
* Fix fetch package vulnerability by @oredavids in https://github.com/gruntwork-io/gruntwork-installer/pull/86

## New Contributors
* @iangrunt made their first contribution in https://github.com/gruntwork-io/gruntwork-installer/pull/75
* @hongil0316 made their first contribution in https://github.com/gruntwork-io/gruntwork-installer/pull/76
* @AlainODea made their first contribution in https://github.com/gruntwork-io/gruntwork-installer/pull/72
* @oredavids made their first contribution in https://github.com/gruntwork-io/gruntwork-installer/pull/86

**Full Changelog**: https://github.com/gruntwork-io/gruntwork-installer/compare/v0.0.38...v0.0.39
<br>

# v0.0.38
_Released Nov 16, 2021_
Updated scripts to support `aarch64`, the alternative name for `arm64`. Thanks to @harrymilne for this contribution! See #70.
<br>

# v0.0.37
_Released Aug 10, 2021_
https://github.com/gruntwork-io/gruntwork-installer/pull/67: Added support for installing `arm64` binaries.
<br>

# v0.0.36
_Released Apr 12, 2021_
#63: Update the help text to include previously missing info on some CLI options.
#64: Clean up a typo in the README.
<br>

# v0.0.35
_Released Mar 3, 2021_
https://github.com/gruntwork-io/gruntwork-installer/pull/60: Expose `--ref` from fetch.
<br>

# v0.0.34
_Released Feb 8, 2021_
#59: Update default fetch version to v0.4.1 which now includes a proper logger.
<br>

# v0.0.33
_Released Feb 4, 2021_
https://github.com/gruntwork-io/gruntwork-installer/pull/58: Update default `fetch` version to the latest, which has a fix for using the `--branch` argument.
<br>

# v0.0.32
_Released Feb 1, 2021_
https://github.com/gruntwork-io/gruntwork-installer/pull/57: Improve the formatting and fix typos in the `--help` text.
<br>

# v0.0.31
_Released Jan 29, 2021_
https://github.com/gruntwork-io/gruntwork-installer/pull/56: Update default `fetch` version. This pulls in fixes/improvements, including support for GitHub Enterprise, skipping Git tags that don't follow SemVer, and paginating on Git tags so that you can fetch older tags in repos with many tags.
<br>

# v0.0.30
_Released Sep 8, 2020_
Just like [v0.0.29](https://github.com/gruntwork-io/gruntwork-installer/releases/tag/v0.0.29), but for branches. The value passed in the `--branch` tag will be exposed as `GRUNTWORK_INSTALL_BRANCH` to the install script.
<br>

# v0.0.29
_Released Sep 4, 2020_
This release exposes the value provided to `--tag` to any install scripts as an environment variable called `GRUNTWORK_INSTALL_TAG`. Scripts can read this variable to learn which tag has been called.
<br>

# v0.0.28
_Released May 20, 2020_
https://github.com/gruntwork-io/gruntwork-installer/pull/49: You can now tell `gruntwork-install` to avoid using `sudo` when installing binaries (NOTE this does not work for module installs) by passing in the arg `--no-sudo "true"`.
<br>

# v0.0.27
_Released May 8, 2020_
https://github.com/gruntwork-io/gruntwork-installer/pull/50 : Minor fixes to ensure scripts pass [shellcheck](https://www.shellcheck.net/).
<br>

# v0.0.26
_Released May 5, 2020_
https://github.com/gruntwork-io/gruntwork-installer/pull/44: Fix bugs in how `gruntwork-install` parses, passes around, and logs `--module-param` arguments to ensure whitespace is handled correctly.

Special thanks to @jeckhart for the PR!
<br>

# v0.0.25
_Released May 4, 2020_
https://github.com/gruntwork-io/gruntwork-installer/pull/46, https://github.com/gruntwork-io/gruntwork-installer/pull/47: Switch from `local readonly` to `local -r`, which is the proper way to declare local, read-only variables in Bash. 

Special thanks to @jeckhart for the fix!
<br>

# v0.0.24
_Released Jan 20, 2020_
#39 Update CodeOwners
#40 Fix broken links in readme
#41 Use a newer terragrunt version in test
<br>

# v0.0.23
_Released Aug 8, 2019_
https://github.com/gruntwork-io/gruntwork-installer/pull/34: Allow omitting github token environment variable when accessing public repo.
<br>

# v0.0.22
_Released Nov 20, 2018_
https://github.com/gruntwork-io/gruntwork-installer/pull/32: Update to fetch v0.3.2, which includes several improvements, including support for GitHub Enterprise.
<br>

# v0.0.21
_Released Mar 1, 2018_
- `gruntwork-install` now uses [v0.2.0 of fetch](https://github.com/gruntwork-io/fetch/releases/tag/v0.2.0), and supports two new parameters: ` --binary-sha256-checksum` and ` --binary-sha512-checksum`, which can be used to validate the binary asset being installed with `gruntwork-install`.
<br>

# v0.0.20
_Released Nov 16, 2017_
https://github.com/gruntwork-io/gruntwork-installer/pull/20: The bootstrap script now supports 32-bit, 686 architecture.
<br>

# v0.0.19
_Released Nov 14, 2017_
- #18: Fixed an issue where `--module-param` passed arguments surrounded by single quotes as a literal (e.g. `'value'` instead of `value`).
- #19: Add better log formatting.

test/no-sudo-test.sh
#!/bin/bash
#
# Some basic automated tests for gruntwork-installer

set -e

readonly LOCAL_INSTALL_URL="file:///src/gruntwork-install"

echo "Using local copy of bootstrap installer to install local copy of gruntwork-install"
./src/bootstrap-gruntwork-installer.sh --download-url "$LOCAL_INSTALL_URL" --version "ignored-for-local-install" --no-sudo "true"

echo "Using gruntwork-install to install a binary from the gruntkms repo into a different folder without using sudo"
gruntwork-install \
  --binary-name "gruntkms" \
  --repo "https://github.com/terraform-modules-krish/gruntkms" \
  --tag "v0.0.1" \
  --binary-install-dir "$HOME" \
  --no-sudo "true"

echo "Checking that gruntkms installed correctly into home dir"
"$HOME/gruntkms" --help


test/integration-test.sh
#!/bin/bash
#
# Some basic automated tests for gruntwork-installer

set -e

readonly LOCAL_INSTALL_URL="file:///src/gruntwork-install"
readonly SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"

echo "Using local copy of bootstrap installer to install local copy of gruntwork-install"
./src/bootstrap-gruntwork-installer.sh --download-url "$LOCAL_INSTALL_URL" --version "ignored-for-local-install"

echo "Using gruntwork-install to install a module from the terraform-aws-ecs repo using branch"
gruntwork-install --module-name "ecs-scripts" --repo "https://github.com/terraform-modules-krish/terraform-aws-ecs" --branch "v0.0.1"

echo "Using gruntwork-install to install a module from the terraform-aws-ecs repo with --download-dir option"
gruntwork-install --module-name "ecs-scripts" --repo "https://github.com/terraform-modules-krish/terraform-aws-ecs" --branch "v0.0.1" --download-dir ~/tmp

echo "Checking that the ecs-scripts installed correctly"
configure-ecs-instance --help

echo "Using gruntwork-install to install a module from the gruntwork-install repo and passing args to it via --module-param"
gruntwork-install --module-name "dummy-module" --repo "https://github.com/terraform-modules-krish/gruntwork-installer" --tag "v0.0.25" --module-param "file-to-cat=$SCRIPT_DIR/integration-test.sh"

echo "Using gruntwork-install to install a module from the gruntwork-install repo with branch as ref"
gruntwork-install --module-name "dummy-module" --repo "https://github.com/terraform-modules-krish/gruntwork-installer" --ref "for-testing-dont-delete" --module-param "file-to-cat=$SCRIPT_DIR/integration-test.sh"

echo "Using gruntwork-install to install a module from the gruntwork-install repo with tag as ref"
gruntwork-install --module-name "dummy-module" --repo "https://github.com/terraform-modules-krish/gruntwork-installer" --ref "v0.0.25" --module-param "file-to-cat=$SCRIPT_DIR/integration-test.sh"

echo "Using gruntwork-install to install a test module from the gruntwork-install repo and test that it's args are maintained via --module-param"
gruntwork-install --module-name "args-test" --repo "https://github.com/terraform-modules-krish/gruntwork-installer" --tag "v0.0.25" --module-param 'test-args=1 2 3 *'

echo "Using gruntwork-install to install a binary from the gruntkms repo"
gruntwork-install --binary-name "gruntkms" --repo "https://github.com/terraform-modules-krish/gruntkms" --tag "v0.0.1"

echo "Checking that gruntkms installed correctly"
gruntkms --help

echo "Unsetting GITHUB_OAUTH_TOKEN to test installing from public repo (terragrunt)"
unset GITHUB_OAUTH_TOKEN

echo "Verifying private repo access is denied"
if gruntwork-install --binary-name "gruntkms" --repo "https://github.com/terraform-modules-krish/gruntkms" --tag "v0.0.1" ; then
  echo "ERROR: was able to access private repo"
  exit 1
fi

echo "Verifying public repo access is allowed"
gruntwork-install --repo 'https://github.com/terraform-modules-krish/terragrunt' --binary-name terragrunt --tag '~>v0.21.0'

echo "Checking that terragrunt installed correctly"
terragrunt --help


bootstrap-gruntwork-installer.sh
#!/bin/bash
#
# A bootstrap script to install the Gruntwork Installer.
#
# Why:
#
# The goal of the Gruntwork Installer is to make make installing Gruntwork Script Modules feel as easy as installing a
# package using apt-get, brew, or yum. However, something has to install the Gruntwork Installer first. One option is
# for each Gruntwork client to do so manually, which would basically entail copying and pasting all the code below.
# This is tedious and would give us no good way to push updates to this bootstrap script.
#
# So instead, we recommend that clients use this tiny bootstrap script as a one-liner:
#
# curl -LsS https://raw.githubusercontent.com/terraform-modules-krish/gruntwork-installer/master/bootstrap-gruntwork-installer.sh | bash /dev/stdin --version 0.0.3
#
# You can copy this one-liner into your Packer and Docker templates and immediately after, start using the
# gruntwork-install command.

set -e

readonly BIN_DIR="/usr/local/bin"
readonly USER_DATA_DIR="/etc/user-data"

readonly DEFAULT_FETCH_VERSION="v0.4.2"
readonly FETCH_DOWNLOAD_URL_BASE="https://github.com/gruntwork-io/fetch/releases/download"
readonly FETCH_INSTALL_PATH="$BIN_DIR/fetch"

readonly GRUNTWORK_INSTALLER_DOWNLOAD_URL_BASE="https://raw.githubusercontent.com/terraform-modules-krish/gruntwork-installer//"
readonly GRUNTWORK_INSTALLER_INSTALL_PATH="$BIN_DIR/gruntwork-install"
readonly GRUNTWORK_INSTALLER_SCRIPT_NAME="gruntwork-install"

function print_usage {
  echo
  echo "Usage: bootstrap-gruntwork-installer.sh [OPTIONS]"
  echo
  echo "A bootstrap script to install the Gruntwork Installer ($GRUNTWORK_INSTALLER_SCRIPT_NAME)."
  echo
  echo "Options:"
  echo
  echo -e "  --version\t\tRequired. The version of $GRUNTWORK_INSTALLER_SCRIPT_NAME to install (e.g. 0.0.3)."
  echo -e "  --fetch-version\tOptional. The version of fetch to install. Default: $DEFAULT_FETCH_VERSION."
  echo -e "  --user-data-owner\tOptional. The user who shown own the $USER_DATA_DIR folder. Default: (current user)."
  echo -e "  --download-url\tOptional. The URL from where to download $GRUNTWORK_INSTALLER_SCRIPT_NAME. Mostly used for automated tests. Default: $GRUNTWORK_INSTALLER_DOWNLOAD_URL_BASE/(version)/$GRUNTWORK_INSTALLER_SCRIPT_NAME."
  echo -e "  --no-sudo\tOptional. When true, don't use sudo to install binaries. Default: false"
  echo
  echo "Examples:"
  echo
  echo "  Install version 0.0.3:"
  echo "    bootstrap-gruntwork-installer.sh --version 0.0.3"
  echo
  echo "  One-liner to download this bootstrap script from GitHub and run it to install version 0.0.3:"
  echo "    curl -Ls https://raw.githubusercontent.com/terraform-modules-krish/gruntwork-installer/master/bootstrap-gruntwork-installer.sh | bash /dev/stdin --version 0.0.3"
}

function maybe_sudo {
  local -r no_sudo="$1"
  shift

  if [[ "$no_sudo" == "true" ]]; then
    "$@"
  else
    sudo "$@"
  fi
}

# http://stackoverflow.com/questions/592620/check-if-a-program-exists-from-a-bash-script
function command_exists {
  local -r cmd="$1"
  type "$cmd" > /dev/null 2>&1
}

function download_url_to_file {
  local -r url="$1"
  local -r file="$2"
  local -r tmp_path=$(mktemp "/tmp/gruntwork-bootstrap-download-XXXXXX")
  local -r no_sudo="$3"

  echo "Downloading $url to $tmp_path"
  if command_exists "curl"; then
    local -r status_code=$(curl -L -s -w '%{http_code}' -o "$tmp_path" "$url")
    assert_successful_status_code "$status_code" "$url"

    echo "Moving $tmp_path to $file"
    maybe_sudo "$no_sudo" mv -f "$tmp_path" "$file"
  else
    echo "ERROR: curl is not installed. Cannot download $url."
    exit 1
  fi
}

function assert_successful_status_code {
  local -r status_code="$1"
  local -r url="$2"

  if [[ "$status_code" == "200" ]]; then
    echo "Got expected status code 200"
  elif string_starts_with "$url" "file://" && [[ "$status_code" == "000" ]]; then
    echo "Got expected status code 000 for local file URL"
  else
    echo "ERROR: Expected status code 200 but got $status_code when downloading $url"
    exit 1
  fi
}

function string_starts_with {
  local -r str="$1"
  local -r prefix="$2"

  [[ "$str" == "$prefix"* ]]
}

function string_contains {
  local -r str="$1"
  local -r contains="$2"

  [[ "$str" == *"$contains"* ]]
}

# http://stackoverflow.com/a/2264537/483528
function to_lower_case {
  tr '[:upper:]' '[:lower:]'
}

function get_os_name {
  uname | to_lower_case
}

function get_os_arch {
  uname -m
}

function get_os_arch_gox_format {
  local -r arch=$(get_os_arch)

  if string_contains "$arch" "arm64"; then
    echo "arm64"
  elif string_contains "$arch" "aarch64"; then
    echo "arm64"
  elif string_contains "$arch" "64"; then
    echo "amd64"
  elif string_contains "$arch" "386"; then
    echo "386"
  elif string_contains "$arch" "686"; then
    echo "386" # Not a typo; 686 is also 32-bit and should work with 386 binaries
  elif string_contains "$arch" "arm"; then
    echo "arm"
  fi
}

function download_and_install {
  local -r url="$1"
  local -r install_path="$2"
  local -r no_sudo="$3"

  download_url_to_file "$url" "$install_path" "$no_sudo"
  maybe_sudo "$no_sudo" chmod 0755 "$install_path"
}

function install_fetch {
  local -r install_path="$1"
  local -r version="$2"
  local -r no_sudo="$3"

  local -r os=$(get_os_name)
  local -r os_arch=$(get_os_arch_gox_format)

  if [[ -z "$os_arch" ]]; then
    echo "ERROR: Unrecognized OS architecture: $(get_os_arch)"
    exit 1
  fi

  echo "Installing fetch version $version to $install_path"
  local -r url="${FETCH_DOWNLOAD_URL_BASE}/${version}/fetch_${os}_${os_arch}"
  download_and_install "$url" "$install_path" "$no_sudo"
}

function install_gruntwork_installer {
  local -r install_path="$1"
  local -r version="$2"
  local -r download_url="$3"
  local -r no_sudo="$4"

  echo "Installing $GRUNTWORK_INSTALLER_SCRIPT_NAME version $version to $install_path"
  download_and_install "$download_url" "$install_path" "$no_sudo"
}

function assert_not_empty {
  local -r arg_name="$1"
  local -r arg_value="$2"

  if [[ -z "$arg_value" ]]; then
    echo "ERROR: The value for '$arg_name' cannot be empty"
    print_usage
    exit 1
  fi
}

function create_user_data_folder {
  local -r user_data_folder="$1"
  local -r user_data_folder_owner="$2"
  local -r user_data_folder_readme="$user_data_folder/README.txt"
  local -r no_sudo="$3"

  echo "Creating $user_data_folder as a place to store scripts intended to be run in the User Data of an EC2 instance during boot"
  maybe_sudo "$no_sudo" mkdir -p "$user_data_folder"

maybe_sudo "$no_sudo" tee "$user_data_folder_readme" > /dev/null <<EOF
The /etc/user-data folder contains scripts that should be executed while an EC2 instance is booting as part of its
User Data (http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/user-data.html) configuration.

This folder is not an industry standard, but a convention we use at  Gruntwork (http://www.gruntwork.io) make User Data
scripts manageable. Many of the modules in Script Modules (https://github.com/gruntwork-io/script-modules) need not
only to be installed, but also to execute while a server is booting, and instead of scattering them in random locations
all over the file system, /etc/user-data gives us a single, common place to put them all.
EOF

  maybe_sudo "$no_sudo" chown -R "$user_data_folder_owner" "$user_data_folder"
}

function bootstrap {
  local fetch_version="$DEFAULT_FETCH_VERSION"
  local installer_version=""
  local download_url=""
  local user_data_folder_owner
  user_data_folder_owner="$(id -u -n)"
  local no_sudo="false"

  while [[ $# -gt 0 ]]; do
    local key="$1"

    case "$key" in
      --version)
        installer_version="$2"
        shift
        ;;
      --fetch-version)
        fetch_version="$2"
        shift
        ;;
      --user-data-owner)
        user_data_folder_owner="$2"
        shift
        ;;
      --download-url)
        download_url="$2"
        shift
        ;;
      --no-sudo)
        no_sudo="$2"
        shift
        ;;
      --help)
        print_usage
        exit
        ;;
      *)
        echo "ERROR: Unrecognized option: $key"
        print_usage
        exit 1
        ;;
    esac

    shift
  done

  assert_not_empty "--version" "$installer_version"
  assert_not_empty "--fetch-version" "$fetch_version"
  assert_not_empty "--user-data-owner" "$user_data_folder_owner"

  if [[ -z "$download_url" ]]; then
    download_url="${GRUNTWORK_INSTALLER_DOWNLOAD_URL_BASE}/${installer_version}/${GRUNTWORK_INSTALLER_SCRIPT_NAME}"
  fi

  echo "Installing $GRUNTWORK_INSTALLER_SCRIPT_NAME..."
  install_fetch "$FETCH_INSTALL_PATH" "$fetch_version" "$no_sudo"
  install_gruntwork_installer "$GRUNTWORK_INSTALLER_INSTALL_PATH" "$installer_version" "$download_url" "$no_sudo"
  create_user_data_folder "$USER_DATA_DIR" "$user_data_folder_owner" "$no_sudo"
  echo "Success!"
}

bootstrap "$@"
